### PR TITLE
Add channel shuffle op fp32 + quantized.

### PR DESCRIPTION
**Pull request by kimishpatel**
Summary:
Pytorch does not have native channel shuffle op.
This diff adds that for both fp and quantized tensors.
For FP implementation is inefficient one. For quantized there is a native
QNNPACK op for this.

Test Plan:
buck run caffe2/test:quantization --
quantization.test_quantized.TestQuantizedOps.test_channel_shuffle
X86 implementation for QNNPACK is sse2 so this may not be the most efficient
for x86.

Differential Revision: D20972122

### CODE REVIEW

File: aten/src/ATen/native/ChanelShuffle.cpp [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+#include <algorithm>
+#include <vector>
+
+namespace at {
+namespace native {
+
+Tensor channel_shuffle(const Tensor& self, int64_t groups) {
+  AT_ASSERTM(self.dim() == 4,
+             "channel_shuffle expects 4D input, but got input with sizes ",self.sizes());
```
Comment (dreiss) [inline aten/src/ATen/native/ChanelShuffle.cpp:14]: Maybe relax this constraint to `dim > 2`?  Should be achievable by

1. First calculate num_pixels as the product of all dimensions other than 0 and 1.
1. `input_reshaped = self.reshape({b, groups, oc, num_pixels})`
1. Change `reshape({b, c, h, w})` to `rehsape(self.shape())`

File: aten/src/ATen/native/ChanelShuffle.cpp [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+  int64_t w = self.size(3);
+  AT_ASSERTM((c % groups) == 0,
+             "Number of channels must be divisible gy groups. Got ",
+             c, " channels and ", groups, " groups.");
+  int64_t oc = c / groups;
+
+  auto input_reshaped = self.reshape({b, groups, oc, h, w});
+  return input_reshaped.permute({0 /* b */, 2 /* oc */, 1 /* groups */, 3 /* h */, 4 /* w */})
+                       .contiguous()
+                       .reshape({b, c, h, w});
```
Comment (dreiss) [inline aten/src/ATen/native/ChanelShuffle.cpp:27]: Maybe reshape before calling contiguous, call contiguous with self.suggest_memory_format, and preserve named tensor dimension names?

File: aten/src/ATen/native/quantized/cpu/q_adaavgpool.cpp [hunk 3]
```diff
@@ -227,11 +229,64 @@ Tensor q_adaptive_avg_pool2d(const Tensor& input, IntArrayRef output_size) {
     return output;
   }
 }
+
+#ifdef USE_PYTORCH_QNNPACK
+Tensor qnnpack_adaptive_avg_pool2d(
```
Comment (dreiss) [inline aten/src/ATen/native/quantized/cpu/q_adaavgpool.cpp:234]: This looks like it's in the wrong diff.

File: aten/src/ATen/native/quantized/cpu/qchannel_shuffle.cpp [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+  TORCH_CHECK(
+      self.scalar_type() == kQUInt8,
+      "Quantized channel shuffle works only on uint8_t.",
+      "But got:", self.scalar_type());
+  const Tensor self_nhwc = self.contiguous(MemoryFormat::ChannelsLast);
+  Tensor qy = at::new_qtensor_cpu(
+      self_nhwc.sizes(),
+      TensorOptions(kQUInt8)
+          .device(kCPU)
+          .memory_format(self_nhwc.suggest_memory_format()),
```
Comment (dreiss) [inline aten/src/ATen/native/quantized/cpu/qchannel_shuffle.cpp:36]: I would just hard-code ChannelsLast here.  Up to you.

File: aten/src/ATen/native/quantized/cpu/qchannel_shuffle.cpp [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+  TORCH_CHECK(
+      self.dim() == 4,
+      "channel_shuffle expects 4D input, but got input with sizes ",
+      self.sizes());
+  TORCH_CHECK(
+      self.scalar_type() == kQUInt8,
+      "Quantized channel shuffle works only on uint8_t.",
+      "But got:", self.scalar_type());
+  const Tensor self_nhwc = self.contiguous(MemoryFormat::ChannelsLast);
+  Tensor qy = at::new_qtensor_cpu(
```
Comment (dreiss) [inline aten/src/ATen/native/quantized/cpu/qchannel_shuffle.cpp:32]: Looks like you're just creating qy as a qtensor with the same shape, dtype, device, format, and qparams as self_nhwc.  Maybe there is already a function for this, or maybe there should be?

File: aten/src/ATen/native/quantized/cpu/qchannel_shuffle.cpp [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+      );
+
+  // Degenerate case of just copying.
+  if (groups == 1) {
+    qy.copy_(self_nhwc);
+    return qy.contiguous(self.suggest_memory_format());
+  }
+
+  int64_t channels = self.size(1);
+  TORCH_CHECK((channels % groups) == 0,
```
Comment (dreiss) [inline aten/src/ATen/native/quantized/cpu/qchannel_shuffle.cpp:48]: What happens if groups is negative or zero?  I know that's not valid, but we should give a clear error instead of a bad result or undefined behavior.

File: aten/src/ATen/native/quantized/cpu/qchannel_shuffle.cpp [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+  pytorch_qnnp_operator_t qnnpack_operator{nullptr};
+
+  const pytorch_qnnp_status createStatus = pytorch_qnnp_create_channel_shuffle_nc_x8(
+      groups /* groups */,
+      channels / groups /* group channels */,
+      0 /* flags */,
+      &qnnpack_operator);
+  TORCH_INTERNAL_ASSERT(
+      createStatus == pytorch_qnnp_status_success,
+      "failed to create QNNPACK Add operator");
```
Comment (dreiss) [inline aten/src/ATen/native/quantized/cpu/qchannel_shuffle.cpp:63]: Message is wrong.

File: aten/src/ATen/native/quantized/cpu/qchannel_shuffle.cpp [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+
+  const pytorch_qnnp_status setupStatus = pytorch_qnnp_setup_channel_shuffle_nc_x8(
+      qnnpack_operator,
+      self_nhwc.numel() / channels /* batch size */,
```
Comment (dreiss) [inline aten/src/ATen/native/quantized/cpu/qchannel_shuffle.cpp:67]: Maybe check channels > 0?

File: aten/src/ATen/native/quantized/cpu/qchannel_shuffle.cpp [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+      (uint8_t*)self_nhwc.data_ptr<c10::quint8>() /* slef data */,
```
Comment (dreiss) [inline aten/src/ATen/native/quantized/cpu/qchannel_shuffle.cpp:68]: "slef"?

File: aten/src/ATen/native/quantized/cpu/qchannel_shuffle.cpp [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+      channels /* self stride */,
+      (uint8_t*)qy.data_ptr<c10::quint8>() /* qy data */,
+      channels /* qy stride */);
+  TORCH_INTERNAL_ASSERT(
+      setupStatus == pytorch_qnnp_status_success,
+      "failed to setup QNNPACK Add operator");
+
+#ifdef FBCODE_CAFFE2
```
Comment (dreiss) [inline aten/src/ATen/native/quantized/cpu/qchannel_shuffle.cpp:76]: Can we make mobile_pthreadpool return null in fbcode instead of checking here?

File: aten/src/ATen/native/quantized/cpu/qchannel_shuffle.cpp [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+  const pytorch_qnnp_status setupStatus = pytorch_qnnp_setup_channel_shuffle_nc_x8(
+      qnnpack_operator,
+      self_nhwc.numel() / channels /* batch size */,
+      (uint8_t*)self_nhwc.data_ptr<c10::quint8>() /* slef data */,
+      channels /* self stride */,
+      (uint8_t*)qy.data_ptr<c10::quint8>() /* qy data */,
+      channels /* qy stride */);
+  TORCH_INTERNAL_ASSERT(
+      setupStatus == pytorch_qnnp_status_success,
+      "failed to setup QNNPACK Add operator");
```
Comment (dreiss) [inline aten/src/ATen/native/quantized/cpu/qchannel_shuffle.cpp:74]: Message

File: aten/src/ATen/native/quantized/cpu/qchannel_shuffle.cpp [hunk 12]
```diff
@@ -0,0 +1,116 @@
...
+  const pytorch_qnnp_status runStatus =
+      pytorch_qnnp_run_operator(qnnpack_operator, nullptr /* thread pool */);
+#else
+  pthreadpool_t threadpool = caffe2::mobile_pthreadpool();
+  const pytorch_qnnp_status runStatus =
+      pytorch_qnnp_run_operator(qnnpack_operator, threadpool);
+#endif
+  TORCH_INTERNAL_ASSERT(
+      runStatus == pytorch_qnnp_status_success,
+      "failed to run QNNPACK Add operator");
```
Comment (dreiss) [inline aten/src/ATen/native/quantized/cpu/qchannel_shuffle.cpp:86]: Message

File: aten/src/ATen/native/quantized/cpu/qchannel_shuffle.cpp [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+// Keep the registry in the anonymous namespace.
+namespace {
+class QChannelShuffle final : public c10::OperatorKernel {
+ public:
+  Tensor operator()(Tensor qx, int64_t groups) {
+    return quantized_channel_shuffle_impl(qx, groups);
+  }
+};
+
+static auto registry = c10::RegisterOperators().op(
```
Comment (dreiss) [inline aten/src/ATen/native/quantized/cpu/qchannel_shuffle.cpp:108]: Why do we need bot this and native_functions.yaml?

File: aten/src/ATen/native/quantized/cpu/qconv.cpp [hunk 14]
```diff
@@ -611,14 +613,14 @@ class QConvInt8 final : public c10::OperatorKernel {
         "be greater than 0.")
 
     // Allocate output Tensor and a buffer for QNNPACK to use
-    Tensor output = at::_empty_affine_quantized(
```
Comment (dreiss) [inline aten/src/ATen/native/quantized/cpu/qconv.cpp:614]: Separate diff?

File: torch/nn/modules/channelhuffle.py [hunk 15]
```diff
@@ -0,0 +1,49 @@
+from .module import Module
+from .. import functional as F
+
+
+class ChannelShuffle(Module):
```
Comment (dreiss) [inline torch/nn/modules/channelhuffle.py:5]: Why is this needed?

**Review by dreiss — CHANGES_REQUESTED** [review]

File: aten/src/ATen/native/ChanelShuffle.cpp (hunk 1 above)
Comment (kimishpatel) [reply aten/src/ATen/native/ChanelShuffle.cpp:14]: Sounds reasonable.

File: aten/src/ATen/native/ChanelShuffle.cpp (hunk 2 above)
Comment (kimishpatel) [reply aten/src/ATen/native/ChanelShuffle.cpp:27]: Sure. That probably is doable. Need to look into named tensor stuff though.

File: aten/src/ATen/native/quantized/cpu/q_adaavgpool.cpp (hunk 3 above)
Comment (kimishpatel) [reply aten/src/ATen/native/quantized/cpu/q_adaavgpool.cpp:234]: Yes. This was a stacked diff in phabricator. Does not get exported as such. I needed to do `ghexport`. Ok I will address you comments and then do ghexport for the entire stack. Dont review any other diffs till then.

File: aten/src/ATen/native/quantized/cpu/qchannel_shuffle.cpp (hunk 5 above)
Comment (kimishpatel) [reply aten/src/ATen/native/quantized/cpu/qchannel_shuffle.cpp:32]: Maybe `empty_like`. I can try but I was also trying to avoid dispatch overhead. That was the reason why I replaced `_empty_affine_quantized` in other places with `new_qtensor_cpu`.

File: aten/src/ATen/native/quantized/cpu/qchannel_shuffle.cpp (hunk 4 above)
Comment (kimishpatel) [reply aten/src/ATen/native/quantized/cpu/qchannel_shuffle.cpp:36]: Good idea.

File: aten/src/ATen/native/quantized/cpu/qchannel_shuffle.cpp (hunk 7 above)
Comment (kimishpatel) [reply aten/src/ATen/native/quantized/cpu/qchannel_shuffle.cpp:63]: Thanks for the catch. Will fix.

File: aten/src/ATen/native/quantized/cpu/qchannel_shuffle.cpp (hunk 9 above)
Comment (kimishpatel) [reply aten/src/ATen/native/quantized/cpu/qchannel_shuffle.cpp:68]: Will fix typo.

File: aten/src/ATen/native/quantized/cpu/qchannel_shuffle.cpp (hunk 10 above)
Comment (kimishpatel) [reply aten/src/ATen/native/quantized/cpu/qchannel_shuffle.cpp:76]: Most of this function was just lifted from C2's implementation. However I think it is cleaner to read this code. However, maybe we should actually pass a valid threadpool. The one that at::parallel_for etc. uses.

File: aten/src/ATen/native/quantized/cpu/qchannel_shuffle.cpp (hunk 13 above)
Comment (kimishpatel) [reply aten/src/ATen/native/quantized/cpu/qchannel_shuffle.cpp:108]: Right. We dont quite need this. I followed the convention for some other quantized ops. But we dont strictly need this. I think the other ops probably access them directly in their unit tests. I will remove this.

File: aten/src/ATen/native/quantized/cpu/qconv.cpp (hunk 14 above)
Comment (kimishpatel) [reply aten/src/ATen/native/quantized/cpu/qconv.cpp:614]: Should have used `ghexport`. Will do that.

File: torch/nn/modules/channelhuffle.py (hunk 15 above)
Comment (kimishpatel) [reply torch/nn/modules/channelhuffle.py:5]: So pixel_shuffle has a similar nn module, which is what I followed. But many times model are written with like `self.channel_shuffle = nn.ChannelShuffle(groups)`. And then their forward just call `self.channel_shuffle(input)`. This is what I have seen. So this maintains that convention.

File: aten/src/ATen/native/quantized/cpu/qchannel_shuffle.cpp (hunk 10 above)
Comment (supriyar) [reply aten/src/ATen/native/quantized/cpu/qchannel_shuffle.cpp:76]: I don't think this ifdef is necessary. Please check other files like qadd.cpp

File: aten/src/ATen/native/quantized/cpu/qchannel_shuffle.cpp [hunk 16]
```diff
@@ -0,0 +1,116 @@
...
+      groups /* groups */,
+      channels / groups /* group channels */,
+      0 /* flags */,
+      &qnnpack_operator);
+  TORCH_INTERNAL_ASSERT(
+      createStatus == pytorch_qnnp_status_success,
+      "failed to create QNNPACK Add operator");
+
+  const pytorch_qnnp_status setupStatus = pytorch_qnnp_setup_channel_shuffle_nc_x8(
+      qnnpack_operator,
```
Comment (supriyar) [inline aten/src/ATen/native/quantized/cpu/qchannel_shuffle.cpp:66]: You'll need to make sure this gets deleted.

File: aten/src/ATen/native/quantized/cpu/qchannel_shuffle.cpp [hunk 17]
```diff
@@ -0,0 +1,116 @@
...
+class QChannelShuffle final : public c10::OperatorKernel {
+ public:
+  Tensor operator()(Tensor qx, int64_t groups) {
+    return quantized_channel_shuffle_impl(qx, groups);
+  }
+};
+
+static auto registry = c10::RegisterOperators().op(
+    "quantized::channel_shuffle(Tensor qx, int groups) -> Tensor qy",
+    c10::RegisterOperators::options()
```
Comment (supriyar) [inline aten/src/ATen/native/quantized/cpu/qchannel_shuffle.cpp:110]: I think there is a change now in the op registration API - https://github.com/pytorch/pytorch/pull/36531

File: torch/nn/modules/channelhuffle.py [hunk 18]
```diff
@@ -0,0 +1,49 @@
+from .module import Module
```
Comment (supriyar) [inline torch/nn/modules/channelhuffle.py:1]: File name has a typo

File: torch/nn/functional.py [hunk 19]
```diff
@@ -2764,6 +2764,44 @@ def multi_margin_loss(input, target, p=1, margin=1., weight=None, size_average=N
...
+channel_shuffle = _add_docstr(torch.channel_shuffle, r"""
+Divide the channels in a tensor of shape :math:`(*, C , H, W)`
+into g groups and rearrange them as :math:`(*, C \frac g, g, H, W)`,
+while keeping the original tensor shape.
+
+See :class:`~torch.nn.ChannelShuffle` for details.
+
+Args:
+    input (Tensor): the input tensor
+    gorups (int): number of groups to divide channels in and rearrange.
```
Comment (supriyar) [inline torch/nn/functional.py:2776]: typo - groups

File: aten/src/ATen/native/quantized/cpu/qchannel_shuffle.cpp (hunk 16 above)
Comment (kimishpatel) [reply aten/src/ATen/native/quantized/cpu/qchannel_shuffle.cpp:66]: Oh yes. Thanks for the catch.

File: aten/src/ATen/native/ChanelShuffle.cpp (hunk 2 above)
Comment (kimishpatel) [reply aten/src/ATen/native/ChanelShuffle.cpp:27]: @dreiss, thinking about this further, contiguous needs to be called before, since we want the groups contiguous in memory. And this very likely restricts us using NHWC format as well. I am gonna leave some comments to the effect.
XNNPACK does have channel shuffle op so for NHWC we can use that even for non-mobile usecase, until non-mobile use case has a better implementation.

File: aten/src/ATen/native/quantized/cpu/qchannel_shuffle.cpp (hunk 5 above)
Comment (kimishpatel) [reply aten/src/ATen/native/quantized/cpu/qchannel_shuffle.cpp:32]: I am gonna leave it as is, due to dispatch overhead concerns.

File: aten/src/ATen/native/quantized/cpu/qchannel_shuffle.cpp (hunk 10 above)
Comment (kimishpatel) [reply aten/src/ATen/native/quantized/cpu/qchannel_shuffle.cpp:76]: OK, looks like to other implementations, such as qadd etc dont have this ifdef. I still think it would have been cleaner if to use mobile_threadpool only for mobile. Now we have code where we are using mobile threadpool even when we may not be running on mobile usecase. It makes this little difficult to read. But for consistency sake, I will stick with the convention.

File: aten/src/ATen/native/quantized/cpu/qchannel_shuffle.cpp (hunk 16 above)
Comment (kimishpatel) [reply aten/src/ATen/native/quantized/cpu/qchannel_shuffle.cpp:66]: BTW, existing implementations do it this way: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/cpu/qadd.cpp#L182-L183. 
I am not sure why is this better than calling the deleter directly at the end. We are just constructing a unique_ptr in the same scope where we expect it to be deleted. Dont think it is much of an overhead, so I will stick with the convention for now. Maybe it is cleaner.

File: torch/nn/modules/channelhuffle.py (hunk 18 above)
Comment (kimishpatel) [reply torch/nn/modules/channelhuffle.py:1]: Good catch.

**Comment by kimishpatel** [comment]
@dreiss, @supriyar, I will close this PR. Created a new stack with ghexport flow which shows stacked diffs correctly. Here: https://github.com/pytorch/pytorch/pull/36815. Also do remember to accept it, eventually, on phabricator since this set of diffs started there.

File: aten/src/ATen/native/quantized/cpu/qchannel_shuffle.cpp (hunk 16 above)
Comment (dreiss) [reply aten/src/ATen/native/quantized/cpu/qchannel_shuffle.cpp:66]: `unique_ptr` ensures that it gets freed even if you throw an exception.
